blueprints/flexible-topologies.md

```diff
@@ -0,0 +1,395 @@
+specific language governing permissions and limitations
+under the License.
+-->
+
+# Flexible Topologies
+
+## Problem Description
+
+Today, a Traffic Control CDN is limited to 2 tiers -- *EDGE* and *MID* -- with the option to skip the *MID* tier for certain Delivery Service types (e.g. `HTTP_LIVE` and `HTTP_NO_CACHE`). In addition, a CDN is limited to one global parent hierarchy, which is defined via the `parent_cachegroup` and `secondary_parent_cachegroup` fields of cachegroups. Both of these problems limit a CDN's ability to scale with increased demand and changing usage patterns, and providing the ability to add more tiers to a CDN helps it keep up with that growth. A Topology that works well for one set of Delivery Services might not be ideal for a different set of Delivery Services, and a CDN needs the flexibility to provide the best Topology for any given Delivery Service -- with any number of tiers and custom caching hierarchies.
+
+## Proposed Change
+
+Traffic Control will provide the ability to define one or more Topologies, and a Topology can have any number of Delivery Services assigned to it. A Topology will be composed of Cachegroups along with their primary/secondary parent relationships to other Cachegroups as defined by the Topology.
+
+If a Delivery Service is assigned to a Topology, any `deliveryservice_server` assignments it has to `EDGE` caches will be ignored, because it will be assigned to all caches in the Delivery Service's CDN (filtered by server capabilities) that belong to the Topology's cachegroups. Ideally, this feature will obsolete legacy `deliveryservice_server` assignments, since Topologies negate the need to assign Delivery Services to individual `EDGE` caches. Nonetheless, legacy `deliveryservice_server` assignments will be supported alongside Topology-based Delivery Services for some time until all Delivery Services have been migrated to Topologies.
+
+### Traffic Portal Impact
+
+Traffic Portal will need new pages for creating and viewing Topologies, and the Delivery Service form will need to be updated to add a new Topology field for assigning a Delivery Service to a Topology. If a Delivery Service is assigned to a Topology, Traffic Portal should prohibit assigning `EDGE` servers to the Delivery Service (`ORIGIN` servers may still need to be assignable for MSO).
+
+Since Delivery Services will no longer be constrained to one global Topology as they are today, it would be extremely useful to be able to visualize a Delivery Service's Topology like a tree, where each node in the tree is a cachegroup, and the edges between nodes are the primary/secondary parent relationships between them. Clicking on a particular node would show all the servers in that cachegroup that could serve a request for the Delivery Service. This visualization will most likely be different from the Topology form for creating a Topology and does not necessarily need to be provided by Traffic Portal.
```

Comment:
`Clicking on a particular node would show all the servers in that cachegroup that could serve a request for the Delivery Service`

This probably requires the following API support:

GET /servers?cg=x&cdn=y&ds=z so servers can be filtered by cg, then cdn (as cachegroups can have servers of multiple cdns), then ds (as a ds's required server capabilities are yet another filter. However, I suppose some of this filtering can be done on the UI side....but not sure about the server capability filter

GET /servers?cachegroup=mid-west&cdn=cdn1&

Reply:
fwiw all of that filtering could be done client-side, but I'm not against adding query params either

Reply:
i think supporting /servers?topology=x should probably be fine and let the UI filter out from there if needed.

Reply:
> i think supporting /servers?topology=x should probably be fine and let the UI filter out from there if needed.

actually i don't think that will work. the server object does not have server capabilities attached to it so it's impossible to filter out servers that don't have the capabilities required of the DS. Maybe this endpoint needs to support the topology query param:

GET deliveryservices/{id}/servers/eligible?topology=foo

^^ deliveryservices/{id}/servers/eligible already removes servers not in a ds's cdn / that don't have the ds's required capabilities.